arbor/fvm_layout.cpp

```diff
@@ -190,69 +201,33 @@ cv_geometry cv_geometry_from_ends(const cable_cell& cell, const locset& lset) {
             }
         }
 
-        auto empty_cable = [](mcable c) { return c.prox_pos==c.dist_pos; };
-        cables.erase(std::remove_if(cables.begin(), cables.end(), empty_cable), cables.end());
-
-        if (!cables.empty()) {
-            if (++head_count[origin(cables.front())]==2) {
-                ++extra_cv_count;
-            }
-
-            n_cv_cables.push_back(cables.size());
-            sort(cables);
-            append(all_cables, std::move(cables));
-        }
-    }
-
-    geom.cv_cables.reserve(all_cables.size()+extra_cv_count);
-    geom.cv_parent.reserve(n_cv_cables.size()+extra_cv_count);
-    geom.cv_cables_divs.reserve(n_cv_cables.size()+extra_cv_count+1);
-    geom.cv_cables_divs.push_back(0);
-
-    mlocation_map parent_map;
-
-    unsigned all_cables_index = 0;
-    unsigned cv_index = 0;
-
-    // Multiple CVs meeting at (0,0)?
-    mlocation root{0, 0};
-    unsigned n_top_children = value_by_key(head_count, root).value_or(0);
-    if (n_top_children>1) {
-        // Add initial trical CV.
-        geom.cv_parent.push_back(mnpos);
-        geom.cv_cables.push_back(mcable{0, 0, 0});
+        sort(cables);
+        append(geom.cv_cables, std::move(cables));
         geom.cv_cables_divs.push_back(geom.cv_cables.size());
-        parent_map[root] = cv_index++;
+        ++cv_index;
     }
 
-    for (auto n_cables: n_cv_cables) {
-        mlocation head = origin(all_cables[all_cables_index]);
-        msize_t parent_cv = value_by_key(parent_map, head).value_or(mnpos);
+    auto n_cv = cv_index;
+    arb_assert(n_cv>0);
+    arb_assert(geom.cv_parent.front()==-1);
+    arb_assert(util::all_of(util::subrange_view(geom.cv_parent, 1, n_cv),
+            [](auto v) { return v!=no_parent; }));
 
-        auto cables = util::subrange_view(all_cables, all_cables_index, all_cables_index+n_cables);
-        std::copy(cables.begin(), cables.end(), std::back_inserter(geom.cv_cables));
+    // Construct CV children mapping by sorting CV indices by parent.
+    util::assign(geom.cv_children, util::make_span(1, n_cv));
+    util::sort_by(geom.cv_children, [&geom](auto cv) { return geom.cv_parent[cv]; });
```

Comment:
What if two CVs have the same parent index? Does the order of their indexes in the sorted list matter?

Reply:
It's definitely relying upon the sort putting CV indices together which have the same parent. There is no guarantee of sorting within the CV child list — but only because that wasn't in my use case. Making them sorted is just a matter of using a stable sort instead, and the extra cost is small.

Reply:
If the only requirement is that they are sorted by parent, no problem. I was just making sure that this wouldn't cause subtle bugs, or different results if the sort algorithm gives a different result.